File: algorithms/sorting/selection-sort/implementation.py
"""Selection Sort - O(n²) time, O(1) space"""

from typing import List

def selection_sort(arr: List[int]) -> List[int]:
    """
    >>> selection_sort([64, 25, 12, 22, 11])
    [11, 12, 22, 25, 64]
    >>> selection_sort([1])
    [1]
    >>> selection_sort([])
    []
    >>> selection_sort([5, 4, 3, 2, 1])
    [1, 2, 3, 4, 5]
    """
    for i in range(len(arr) - 1):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

if __name__ == '__main__':
    import doctest
    r = doctest.testmod()
    print(f'✓ {r.attempted} doctests passed!' if r.failed == 0 else f'✗ {r.failed} failed')

File: algorithms/sorting/selection-sort/implementation.js
/**
 * Selection Sort - repeatedly select minimum and swap
 * Time: O(n²), Space: O(1), Stable: No
 */

function selectionSort(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        let minIdx = i;
        for (let j = i + 1; j < arr.length; j++) {
            if (arr[j] < arr[minIdx]) minIdx = j;
        }
        if (minIdx !== i) {
            [arr[i], arr[minIdx]] = [arr[minIdx], arr[i]];
        }
    }
    return arr;
}

function selectionSortDescending(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        let maxIdx = i;
        for (let j = i + 1; j < arr.length; j++) {
            if (arr[j] > arr[maxIdx]) maxIdx = j;
        }
        if (maxIdx !== i) {
            [arr[i], arr[maxIdx]] = [arr[maxIdx], arr[i]];
        }
    }
    return arr;
}

if (require.main === module) {
    console.log('Selection Sort:', selectionSort([64, 25, 12, 22, 11]));
    console.log('Descending:', selectionSortDescending([64, 25, 12, 22, 11]));
}

module.exports = { selectionSort, selectionSortDescending };
